src/tool/hpcrun/gpu/gpu-metrics.c

```diff
@@ -583,6 +606,16 @@ gpu_metrics_default_enable
 
   FINALIZE_METRIC_KIND();
 
+// Execution count metrics
+#undef CURRENT_METRIC 
+#define CURRENT_METRIC GCOUNT
+
+  INITIALIZE_METRIC_KIND();
+
+  FORALL_GCOUNT(INITIALIZE_SCALAR_METRIC_INT)
```

Comment:
There's a design issue.

All metrics, including GMEM, GMSET, GXCOPY, and GICOPY are `INDEXED_METRIC`, while to increase GMEM_COUNT, GMSET_COUNT, GXCOPY_COUNT, and GICOPY_COUNT, we need `SCALAR_METRIC`. 

Thus, I do not recommend to add COUNT as an index in existing GPU API metrics. Rather, we shall keep the current solution that uses a single GCOUNT metric, or init GMEM_COUNT, GMSET_COUNT, GXCOPY_COUNT, and GICOPY_COUNT separately as different metrics.

Reply:
Adding @mxz297 to the conversation:

Regarding GMEM: 

Without correlation ID, we can't properly attribute them if we use the activity API. Can't we attribute them using the callback API? (I think this would be a change from the implementation now, but I haven't checked)

Regarding using COUNT with INDEXED_METRICs:

If we use type indices 0..N, why is it a bad idea to record the count using index N+1? Alternatively, we can use type indices as 1..N+1 and record the count as index 0. 
Why do I propose this? I think keeping all of the relevant things together is useful. If we monitor one of the types, we will monitor all of them (count included).

I don't know whether roctracer has correlation ids for allocation and deallocatiion.

Reply:
> Without correlation ID, we can't properly attribute them if we use the activity API. Can't we attribute them using the callback API? (I think this would be a change from the implementation now, but I haven't checked)

That needs a lot of implementation changes. Inside `cupti-api.c`, we have to manually identify every memory allocation API, read parameters, and calculate the time spent in these APIs. I suggest to add it in the next version.

> If we use type indices 0..N, why is it a bad idea to record the count using index N+1? Alternatively, we can use type indices as 1..N+1 and record the count as index 0.

Yes, we could. My thought was that the INDEXED_METRICs are used for activity **types**, reserving the last slot for COUNTs might be confusing for future development. If we decide to do so, we would better add some comments.

Reply:
Another problem of adding COUNT as the last index.

> FORALL_GSYNC(INITIALIZE_INDEXED_METRIC_REAL)

GSYNC metrics are REAL but not INT, adding 1 to a REAL metric does not have effect, though theoretically we could add 1.0 to GSYNC_COUNT.

Reply:
It would be OK to add 1.0 to counts as a real.